optimum/fx/optimization/transformations.py

```diff
@@ -803,3 +799,26 @@ def reverse(self, graph_module):
                 return ComposeTransformation._reverse_composition(graph_module)
 
     return ComposeTransformation()
+
+
+# was removed from transformers in favor of gen_constructor_wrapper which only works during tracing
+# TODO: should transformation be applied during tracing ? or by a torch.compile backend instead ?
+def _gen_constructor_wrapper(target):
```

Comment:
do you think we can add `gen_constructor_wrapper` https://github.com/huggingface/transformers/blob/87a351818e6c680282d86a00b61fe4c55b41477d/src/transformers/utils/fx.py#L795 directly after v4.40.2 (included) ? cc @IlyasMoutawwakil @michaelbenayoun

Reply:
I went with copying the old function but apparently it's possible to force tracing mode with:
`torch.fx._symbolic_trace._is_fx_tracing_flag = True`
we can use both: 
- `_gen_constructor_wrapper` from transformers when it's available, 
- and a new one defined as:
```python
def _gen_constructor_wrapper(*args, **kwargs):
    original_flag = torch.fx._symbolic_trace._is_fx_tracing_flag
    torch.fx._symbolic_trace._is_fx_tracing_flag = True
    out = gen_constructor_wrapper(*args, **kwargs)
    torch.fx._symbolic_trace._is_fx_tracing_flag = original_flag 
    return out
```

Reply:
I think the above solution might not work no?
Because the test if we are tracing is inside the wrapper itself, so it needs to `is_fx_tracing` needs to be `True` when the wrapper is called.

Reply:
I tested it and it works
````
import torch.fx._symbolic_trace
from torch.fx._symbolic_trace import is_fx_tracing

torch.fx._symbolic_trace._is_fx_tracing_flag = True
print(is_fx_tracing())
torch.fx._symbolic_trace._is_fx_tracing_flag = False
print(is_fx_tracing())
```
Returns True then False, ie inside the wrapper when `is_fx_tracing` will be called, it will return the flag value that we forced. I will implement the solution and test with the two versions of transformers.

Reply:
I pushed a fix that wraps the wrapper output of `gen_constructor_wrapper`